src/utils/daos.c

```diff
@@ -0,0 +1,870 @@
+		fprintf(stderr, "pool UUID required\n");
+		fflush(stderr);
+		return RC_PRINT_HELP;
+	}
+
+	/* Check the pool service ranks.(required) */
+	if (ap->mdsrv_str == NULL) {
+		fprintf(stderr, "--svc must be specified\n");
+		fflush(stderr);
+		return RC_PRINT_HELP;
+	}
+
+	if (ap->mdsrv == NULL) {
+		fprintf(stderr, "failed to parse --svc=%s\n", ap->mdsrv_str);
+		return RC_NO_HELP;
+	}
+
+	if (ap->mdsrv->rl_nr == 0) {
+		fprintf(stderr, "--svc must not be empty\n");
+		fflush(stderr);
+		rc = RC_PRINT_HELP;
+		goto bad_opt_free_mdsrv;
+	}
+
+	/* TODO: decide if for container operations the code should
+	 * check that a container UUID or path has been provided,
+	 * and if a path, that container UUID can be looked up (UNS).
+	 * Or, if that checking should be in container command handling.
+	 */
+	return 0;
+
+bad_opt_free_mdsrv:
+	daos_rank_list_free(ap->mdsrv);
+	return rc;
+}
+
+/* For operations that take <pool_uuid, pool_group, pool_svc_ranks>. */
+static int
+pool_op_hdlr(struct cmd_args_s *ap)
+{
+	daos_handle_t		pool;
+	int			rc = 0;
+	int			rc2;
+	enum pool_op		op;
+
+	assert(ap != NULL);
+	op = ap->p_op;
+
+	/* TODO: create functions per pool op */
+
+	if (op == POOL_QUERY) {
+		daos_pool_info_t		 pinfo;
+		struct daos_pool_space		*ps = &pinfo.pi_space;
+		struct daos_rebuild_status	*rstat = &pinfo.pi_rebuild_st;
+		int				 i;
+
+		rc = daos_pool_connect(ap->pool_uuid, ap->group,
+				       ap->mdsrv, DAOS_PC_RO, &pool,
+				       NULL /* info */, NULL /* ev */);
+		if (rc != 0) {
+			fprintf(stderr, "failed to connect to pool: %d\n", rc);
+			goto bad_pool_connect;
+		}
+
+		pinfo.pi_bits = DPI_ALL;
+		rc = daos_pool_query(pool, NULL, &pinfo, NULL, NULL);
+		if (rc != 0) {
+			fprintf(stderr, "pool query failed: %d\n", rc);
+			goto bad_pool_query;
+		}
+		D_PRINT("Pool "DF_UUIDF", ntarget=%u, disabled=%u\n",
+			DP_UUID(pinfo.pi_uuid), pinfo.pi_ntargets,
+			pinfo.pi_ndisabled);
+
+		D_PRINT("Pool space info:\n");
+		D_PRINT("- Target(VOS) count:%d\n", ps->ps_ntargets);
+		for (i = DAOS_MEDIA_SCM; i < DAOS_MEDIA_MAX; i++) {
+			D_PRINT("- %s:\n",
+				i == DAOS_MEDIA_SCM ? "SCM" : "NVMe");
+			D_PRINT("  Total size: "DF_U64"\n",
+				ps->ps_space.s_total[i]);
+			D_PRINT("  Free: "DF_U64", min:"DF_U64", max:"DF_U64", "
+				"mean:"DF_U64"\n", ps->ps_space.s_free[i],
+				ps->ps_free_min[i], ps->ps_free_max[i],
+				ps->ps_free_mean[i]);
+		}
+
+		if (rstat->rs_errno == 0) {
+			char	*sstr;
+
+			if (rstat->rs_version == 0)
+				sstr = "idle";
+			else if (rstat->rs_done)
+				sstr = "done";
+			else
+				sstr = "busy";
+
+			D_PRINT("Rebuild %s, "DF_U64" objs, "DF_U64" recs\n",
+				sstr, rstat->rs_obj_nr, rstat->rs_rec_nr);
+		} else {
+			D_PRINT("Rebuild failed, rc=%d, status=%d\n",
+				rc, rstat->rs_errno);
+		}
+	}
+
+	/* TODO implement the following:
+	 * op == POOL_LIST_CONTAINERS
+	 * op == POOL_STAT
+	 * op == POOL_GET_PROP
+	 * op == POOL_GET_ATTR
+	 * op == POOL_LIST_ATTRS
+	 */
+
+bad_pool_query:
```

Comment:
i wouldn't call this "bad", since even if everything succeeds above, you still get here.

Reply:
agree, will fix in next patch